### ISSUE TITLE
Behavior of linearRamp started with events far in the past

### ISSUE THREAD

**Issue by rtoy** [WebAudio/web-audio-api#1844]
See #1842, in particular https://github.com/WebAudio/web-audio-api/issues/1842#issuecomment-476878260

Created this new issue to track that.  A copy of the issue:

> I'm planning to add several examples of differences that I'm seeing in the Web Audio behavior in across the popular browsers to this GitHub issue. Please let me know if these should be going somewhere else.

> The first is a difference in the behavior of the linearRampToValueAtTime method. On Chrome, it seems to start the ramp at the time the method is invoked. On Firefox, Chrome, and Edge it seems to calculate the ramp based on the previous event. I'm not sure which behavior is closer to what the spec intended, but I'm finding it problematic when making frequent volume changes and wanting to hear smooth transitions from the current gain setting to the specified value. My guess is that this method was intended more for the creation of amplitude envelopes, and clients could make several calls in a row to define the envelope as a sequence of ramps, but I'm not sure and the online examples don't make this very clear. At any rate, if the behavior were consistent across browsers, it would be easier to figure out how this method can and should be applied.

> The provided example code creates an oscillator and hooks it up to a gain node when the "Start" button is pressed. It then periodically calls linearRampToValueAtTime every two seconds with a ramp time of 1/2 second. On Chrome, smooth level transitions are heard. On Firefox, Safari, and Edge, the volume transitions are abrupt. On some of the browsers in the 2nd list, I'm hearing a bit of residual oscillator sound during the downward ramp of the volume. Pressing the "Stop" button stops and disconnects the oscillator.

> A live link to the test document can be found at https://phet-dev.colorado.edu/jb-tests/periodic-linear-ramp-tester.html (this link should be live for at least a month, no guarantees after that time), and the code can be found at https://github.com/phetsims/tambo/blob/master/html/periodic-linear-ramp-tester.html.

> Please let me know if any additional information is needed.

**Comment by rtoy** [comment]
For a followup comment on this, see https://github.com/WebAudio/web-audio-api/issues/1842#issuecomment-476864921

In summary, Chrome's behavior seems nice, but is wrong according to the spec.

**Comment by karlt** [comment]
I agree that smooth transitions are nice, and I expect that would be a more useful behavior.

**Comment by rtoy** [comment]
Updating the spec to require it to do what chrome does will be hard at this time.  The spec is pretty clear that the slope of the ramp depends on the previous event and these examples have a previous event.

I would prefer that we don't add additional magic to AudioParam timelines.  Developers need to think carefully about scheduling things according to what the spec actually says and not according to what they think it says. (I'm certainly guilty of that!)

**Comment by rtoy** [comment]
Action Item: @rtoy  Describe what's happening here in a bit more detail for the record.  But this is fundamentally a bug in Chrome.

**Comment by rtoy** [comment]
Run the following in https://hoch.github.io/canopy using Chrome (No one else supports context.suspend(t) for an offline context.):
```
// @channels 1
// @duration 1.0
// @sampleRate 44100

var src = new ConstantSourceNode(context);
src.offset.setValueAtTime(0, 0);

context.suspend(.5)
.then(() => {
  src.offset.linearRampToValueAtTime(1, 1);
})
.then(() => context.resume());
src.connect(context.destination);
src.start();
```
What you see is the output is 0 from time 0 to 0.5.  Then there's a ramp going from 0 to 1 from time 0.5 to 1.

The spec is pretty clear here.  The ramp should start from the previous event which is the `setValueAtTime` so the curve should be a ramp starting at value 0 at time 0 to 1 at time 1.  But the ramp is inserted at time 0.5, so the output should be a jump from 0 to 0.5 and then a linear ramp up to 1.

This is a bug in Chrome.

**Comment by meszaros-lajos-gyorgy** [comment]
So if I get it right, we should always do an `AudioParam.setValueAtTime(currentValue, ctx.currentTime)` before starting a new ramp?

**Comment by rtoy** [comment]
No.  Unless you really want that.  The ramp is supposed to start off where the previous event ended.  (If it did end; setTarget never really ends.)

Is this an issue with the spec? I think it's pretty clear about how ramps should behave. If it is an issue with the spec, please file a new bug.
